modules/basics/src/main/java/com/opengamma/basics/currency/FxMatrix.java

```diff
@@ -0,0 +1,557 @@
+   */
+  public static FxMatrix.Builder builder() {
+    return new FxMatrix.Builder();
+  }
+
+  /**
+   * Create a new builder using the data from this matrix to
+   * create a set of initial entries.
+   *
+   * @return a new builder containing the data from this matrix
+   */
+  public FxMatrix.Builder toBuilder() {
+    return new FxMatrix.Builder(currencies, rates);
+  }
+
+  /**
+   * Return the exchange rate between two currencies.
+   *
+   * @param ccy1 The first currency.
+   * @param ccy2 The second currency.
+   * @return The exchange rate: 1.0 * ccy1 = x * ccy2.
+   */
+  public double getRate(Currency ccy1, Currency ccy2) {
+    if (ccy1.equals(ccy2)) {
+      return 1;
+    }
+    Integer index1 = currencies.get(ccy1);
+    Integer index2 = currencies.get(ccy2);
+    if (index1 != null && index2 != null) {
+      return rates[index1][index2];
+    } else {
+      throw new IllegalArgumentException(
+          "No rate found for " + ccy1 + "/" + ccy2 +
+              " - FX matrix only contains rates for: " + currencies.keySet());
+    }
+  }
+
+  /**
+   * Convert a currency amount into a amount in the specified currency
+   * using the rates in this matrix.
+   *
+   * @param amount  the currency amount, not null
+   * @param ccy  the currency to convert all entries to
+   * @return the amount converted to the requested currency
+   */
+  public CurrencyAmount convert(CurrencyAmount amount, Currency ccy) {
+    ArgChecker.notNull(amount, "amount");
+    ArgChecker.notNull(ccy, "ccy");
+    Currency originalCcy = amount.getCurrency();
+    // Only do conversion if we need to
+    return originalCcy == ccy ? amount : CurrencyAmount.of(ccy, amount.getAmount() * getRate(originalCcy, ccy));
+  }
+
+  /**
+   * Convert a multiple currency amount into a amount in the specified currency
+   * using the rates in this matrix.
+   *
+   * @param amount  the multiple currency amount, not null
+   * @param ccy  the currency to convert all entries to
+   * @return the total amount in the requested currency
+   */
+  public CurrencyAmount convert(MultiCurrencyAmount amount, Currency ccy) {
+    ArgChecker.notNull(amount, "amount");
+    ArgChecker.notNull(ccy, "ccy");
+
+    // We could do this using the currency amounts but to
+    // avoid creating extra objects we'll use doubles
+    double total = amount.getAmounts()
+        .stream()
+        .mapToDouble(ca -> ca.getAmount() * getRate(ca.getCurrency(), ccy))
+        .sum();
+    return CurrencyAmount.of(ccy, total);
+  }
+
+  /**
+   * Merge the entries from the other matrix into this one. The other matrix
+   * should have at least one currency in common with this one.
+   * The additional currencies from the other matrix are added one by one
+   * and the exchange rate data created is coherent with some data in this
+   * matrix. If the data in the two matrices are not coherent then there is
+   * no guarantee which data will be used and the final result may be incoherent.
+   *
```

Comment:
Actually, I think it's rubbish - we only add in rates for currencies we don't hold and therefore it will be consistent.